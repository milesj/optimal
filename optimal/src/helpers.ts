import {
	Constructor,
	DefaultValue,
	DefaultValueInitializer,
	Schema,
	SchemaValidateOptions,
	UnknownObject,
} from './types';
import { ValidationError } from './ValidationError';

export function isObject(value: unknown): value is object {
	return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
	return (
		isObject(value) &&
		typeof (value as UnknownObject).schema === 'function' &&
		typeof (value as UnknownObject).state === 'function' &&
		typeof (value as UnknownObject).type === 'function' &&
		typeof (value as UnknownObject).validate === 'function'
	);
}

export function isValidDate(value: Date): value is Date {
	return (
		Object.prototype.toString.call(value) === '[object Date]' &&
		!Number.isNaN(value.getTime()) &&
		value.toString() !== 'Invalid Date'
	);
}

export function isValidNumber(value: unknown): value is number {
	return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
	return typeof value === 'string' && value !== '';
}

export function createArray(value: unknown): unknown[] {
	if (value === undefined) {
		return [];
	}

	return Array.isArray(value) ? [...(value as unknown[])] : [value];
}

export function createDate(value: unknown): Date {
	if (value instanceof Date) {
		return value;
	}

	if (value === undefined || value === null) {
		return new Date();
	}

	return new Date(value as number);
}

export function createObject<T = UnknownObject>(value: unknown) {
	return (isObject(value) ? { ...value } : {}) as unknown as T;
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(object: unknown, contract: Constructor<T>): object is T {
	if (!object || typeof object !== 'object') {
		return false;
	}

	if (object instanceof contract) {
		return true;
	}

	let current = object;

	while (current) {
		if (current.constructor.name === 'Object') {
			return false;
		}

		if (
			current.constructor.name === contract.name ||
			(current instanceof Error && current.name === contract.name)
		) {
			return true;
		}

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		current = Object.getPrototypeOf(current);
	}

	return false;
}

export function invalid(
	condition: boolean,
	message: string,
	path: string = '',
	value: unknown = undefined,
) {
	if (condition) {
		return;
	}

	throw new ValidationError(message, path, value);
}

export function invariant(condition: boolean, message: string) {
	if (condition) {
		return;
	}

	throw new Error(message);
}

export function logUnknown(unknownFields: object, pathPrefix?: string) {
	const unknownKeys = Object.keys(unknownFields);

	if (unknownKeys.length > 0) {
		throw new Error(
			`${pathPrefix ? `Unknown "${pathPrefix}" fields` : 'Unknown fields'}: ${unknownKeys.join(
				', ',
			)}.`,
		);
	}
}

export function pathKey(path: string): string {
	if (path.endsWith(']')) {
		const index = path.lastIndexOf('[');

		return index > 0 ? path.slice(index) : path;
	}

	const index = path.lastIndexOf('.');

	return index > 0 ? path.slice(index + 1) : path;
}

export function prettyValue(value: unknown): string | null {
	switch (typeof value) {
		case 'string':
			return `"${value}"`;

		case 'number':
		case 'function':
			return String(value);

		case 'object': {
			if (value === null) {
				return `\`null\``;
			}

			if (value.constructor !== Object) {
				return value.constructor.name === 'Array' ? null : `\`${value.constructor.name}\``;
			}

			return null;
		}

		default:
			return `\`${value}\``;
	}
}

export function tryAndCollect(
	validator: () => boolean | void,
	validError: ValidationError,
	collectErrors?: boolean,
): boolean {
	let result = false;

	try {
		const value = validator();

		if (typeof value === 'boolean') {
			result = value;
		}
	} catch (error: unknown) {
		if (error instanceof Error && collectErrors) {
			validError.addError(error);
		} else {
			throw error;
		}
	}

	return result;
}

export function extractDefaultValue<T>(
	defaultValue: DefaultValue<T>,
	path: string,
	{ currentObject, rootObject }: SchemaValidateOptions,
) {
	return typeof defaultValue === 'function'
		? (defaultValue as DefaultValueInitializer<T>)(path, currentObject!, rootObject!)
		: defaultValue;
}
