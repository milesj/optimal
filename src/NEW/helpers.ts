import { Constructor, UnknownObject, Schema } from './types';

export function createDate(value: unknown): Date {
  return value instanceof Date ? value : new Date(value as number);
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(
  object: unknown,
  contract: Constructor<T> | Function,
): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof contract) {
    return true;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === 'Object') {
      return false;
    }

    if (
      current.constructor.name === contract.name ||
      (current instanceof Error && current.name === contract.name)
    ) {
      return true;
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    current = Object.getPrototypeOf(current);
  }

  return false;
}

export function invariant(condition: boolean, message: string, path: string = '') {
  if (__DEV__) {
    if (condition) {
      return;
    }

    // TODO no errorPrefix
    const prefix = path ? `Invalid field "${path}".` : '';

    throw new Error(`${prefix} ${message}`.trim());
  }
}

export function isObject(value: unknown): value is object {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
  return (
    isObject(value) &&
    typeof (value as UnknownObject).validate === 'function' &&
    typeof (value as UnknownObject).type === 'function'
  );
}

export function isValidDate(value: Date): value is Date {
  return (
    Object.prototype.toString.call(value) === '[object Date]' &&
    !Number.isNaN(value.getTime()) &&
    value.toString() !== 'Invalid Date'
  );
}

export function isValidNumber(value: unknown): value is number {
  return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
  return typeof value === 'string' && value !== '';
}

export function pathKey(path: string): string {
  const index = path.lastIndexOf('.');

  return index > 0 ? path.slice(index + 1) : path;
}
