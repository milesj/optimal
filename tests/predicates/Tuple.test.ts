import {
  tuple,
  array,
  bool,
  object,
  number,
  string,
  ObjectOf,
  ArrayOf,
  TuplePredicate,
} from '../../src';
import { runChecks, runInProd } from '../helpers';

describe('TuplePredicate', () => {
  type TupleStrings = 'foo' | 'bar' | 'baz';
  type Tuple = [ArrayOf<string>, boolean, number, ObjectOf<number>, TupleStrings];
  let predicate: TuplePredicate<Tuple>;

  beforeEach(() => {
    predicate = tuple([
      array(string()),
      bool(true),
      number(1).between(0, 5),
      object(number()),
      string('foo').oneOf(['foo', 'bar', 'baz']),
    ]);
  });

  it('errors if a non-array is not passed', () => {
    expect(() => {
      // @ts-ignore
      union('foo', []);
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if an empty array is passed', () => {
    expect(() => {
      tuple(
        // @ts-ignore
        [],
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if an array with non-predicates is passed', () => {
    expect(() => {
      tuple([
        // @ts-ignore
        123,
      ]);
    }).toThrowErrorMatchingSnapshot();
  });

  it('doesnt error if a predicate array is passed', () => {
    expect(() => {
      tuple<[string]>([string()]);
    }).not.toThrow();
  });

  describe('run()', () => {
    it('returns an array of default values if undefined provided', () => {
      expect(runChecks(predicate)).toEqual([[], true, 1, {}, 'foo']);
    });

    it('returns an array of values if half the tuple was defined', () => {
      expect(runChecks(predicate, [['a', 'b', 'c'], false])).toEqual([
        ['a', 'b', 'c'],
        false,
        1,
        {},
        'foo',
      ]);
    });

    it('returns an array of all values if defined', () => {
      expect(runChecks(predicate, [['a', 'b', 'c'], false, 3, { a: 1 }, 'bar'])).toEqual([
        ['a', 'b', 'c'],
        false,
        3,
        { a: 1 },
        'bar',
      ]);
    });

    it('errors if too many values defined', () => {
      expect(() => {
        runChecks(
          predicate,
          // @ts-ignore Allow extra
          [['a', 'b', 'c'], false, 3, { a: 1 }, 'bar', 'unknown'],
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs checks for array predicate', () => {
      expect(() => {
        runChecks(
          predicate,
          // @ts-ignore Allow invalid type
          [[123]],
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs checks for boolean predicate', () => {
      expect(() => {
        runChecks(
          predicate,
          // @ts-ignore Allow invalid type
          [['a'], 123],
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs checks for number predicate', () => {
      expect(() => {
        runChecks(predicate, [['a'], true, 10]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs checks for object predicate', () => {
      expect(() => {
        runChecks(
          predicate,
          // @ts-ignore Allow invalid type
          [['a'], true, 3, { a: 'a' }],
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs checks for string predicate', () => {
      expect(() => {
        runChecks(
          predicate,
          // @ts-ignore Allow invalid type
          [['a'], true, 3, {}, 'qux'],
        );
      }).toThrowErrorMatchingSnapshot();
    });

    describe('production', () => {
      it(
        'returns default shape if value is empty',
        runInProd(() => {
          expect(runChecks(predicate, [])).toEqual([[], true, 1, {}, 'foo']);
        }),
      );

      it(
        'returns default shape if value is undefined',
        runInProd(() => {
          expect(runChecks(predicate)).toEqual([[], true, 1, {}, 'foo']);
        }),
      );

      it(
        'bypasses checks and returns value',
        runInProd(() => {
          expect(
            runChecks(
              predicate,
              // @ts-ignore Test invalid type
              [[123], true, 500],
            ),
          ).toEqual([['123'], true, 500, {}, 'foo']);
        }),
      );
    });
  });

  describe('typeAlias()', () => {
    it('returns the type name', () => {
      expect(predicate.typeAlias()).toBe(
        '[array<string>, boolean, number, object<number>, string]',
      );
    });
  });
});
