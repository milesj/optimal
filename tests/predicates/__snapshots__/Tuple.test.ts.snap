// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TuplePredicate errors if a non-array is not passed 1`] = `"union is not defined"`;

exports[`TuplePredicate errors if an array with non-predicates is passed 1`] = `"A non-empty array of blueprints are required for a tuple."`;

exports[`TuplePredicate errors if an empty array is passed 1`] = `"A non-empty array of blueprints are required for a tuple."`;

exports[`TuplePredicate run() errors if too many values defined 1`] = `"Value must be a tuple with less than or equal to 5 items."`;

exports[`TuplePredicate run() runs checks for array predicate 1`] = `"Invalid field \\"key[0][0]\\". Must be a string."`;

exports[`TuplePredicate run() runs checks for boolean predicate 1`] = `"Invalid field \\"key[1]\\". Must be a boolean."`;

exports[`TuplePredicate run() runs checks for number predicate 1`] = `"Invalid field \\"key[2]\\". Number must be between 0 and 5."`;

exports[`TuplePredicate run() runs checks for object predicate 1`] = `"Invalid field \\"key[3].a\\". Must be a number."`;

exports[`TuplePredicate run() runs checks for string predicate 1`] = `"Invalid field \\"key[4]\\". String must be one of: foo, bar, baz"`;
